File: 2641. Cousins in Binary Tree II/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode replaceValueInTree(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        Queue<Pair> pass = new LinkedList<>();
        q.offer(root);
        pass.offer(new Pair(root, root.val));
        int prevSum = 0, prevL = 0;
        while (!pass.isEmpty()) {
            int currSum = 0, l = q.size();
            for (int i = 0; i < l; i++) {
                TreeNode curr = q.poll();
                currSum += curr.val;
                int t = 0;
                if (curr.left != null) t += curr.left.val;
                if (curr.right != null) t += curr.right.val;
                if (curr.left != null) {
                    pass.offer(new Pair(curr.left, t));
                    q.offer(curr.left);
                }
                if (curr.right != null) {
                    pass.offer(new Pair(curr.right, t));
                    q.offer(curr.right);
                }
            }
            for (int i = 0; i < prevL; i++) {
                Pair curr = pass.poll();
                curr.node.val = prevSum - curr.s;    
            }
            prevSum = currSum;
            prevL = l;
        }
        return root;
    }
}
class Pair {
    TreeNode node;
    int s;
    public Pair(TreeNode c, int t) {
        node = c;
        s = t;
    }
}

File: 2641. Cousins in Binary Tree II/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        q, p = [root], [[root, root.val]]
        prevSum, prevL = 0, 0
        while p:
            currSum, l = 0, len(q)
            for i in range(l):
                curr = q.pop(0)
                currSum += curr.val
                t = 0
                if curr.left: t += curr.left.val
                if curr.right: t += curr.right.val
                if curr.left:
                    q.append(curr.left)
                    p.append([curr.left, t])
                if curr.right:
                    q.append(curr.right)
                    p.append([curr.right, t])
            for i in range(prevL):
                curr, t = p.pop(0)
                curr.val = prevSum - t
            prevSum, prevL = currSum, l
        return root